src/lapack/backends/rocsolver/rocsolver_scope_handle.cpp

```diff
@@ -0,0 +1,141 @@
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+#include "rocsolver_scope_handle.hpp"
+#if __has_include(<sycl/detail/common.hpp>)
+#include <sycl/detail/common.hpp>
+#else
+#include <CL/sycl/detail/common.hpp>
+#endif
+
+namespace oneapi {
+namespace mkl {
+namespace lapack {
+namespace rocsolver {
+
+/**
+ * Inserts a new element in the map if its key is unique. This new element
+ * is constructed in place using args as the arguments for the construction
+ * of a value_type (which is an object of a pair type). The insertion only
+ * takes place if no other element in the container has a key equivalent to
+ * the one being emplaced (keys in a map container are unique).
+ */
+thread_local rocsolver_handle<pi_context> RocsolverScopedContextHandler::handle_helper =
+    rocsolver_handle<pi_context>{};
+
+RocsolverScopedContextHandler::RocsolverScopedContextHandler(sycl::queue queue,
+                                                             sycl::interop_handle &ih)
+        : ih(ih),
+          needToRecover_(false) {
+    placedContext_ = new sycl::context(queue.get_context());
+    auto device = queue.get_device();
+    auto desired = sycl::get_native<sycl::backend::ext_oneapi_hip>(*placedContext_);
+    hipError_t err;
+    HIP_ERROR_FUNC(hipCtxGetCurrent, err, &original_);
+    if (original_ != desired) {
+        // Sets the desired context as the active one for the thread
+        HIP_ERROR_FUNC(hipCtxSetCurrent, err, desired);
+        // No context is installed and the suggested context is primary
+        // This is the most common case. We can activate the context in the
+        // thread and leave it there until all the PI context referring to the
+        // same underlying rocblas primary context are destroyed. This emulates
+        // the behaviour of the rocblas runtime api, and avoids costly context
+        // switches. No action is required on this side of the if.
+        needToRecover_ = !(original_ == nullptr);
+    }
+}
+
+RocsolverScopedContextHandler::~RocsolverScopedContextHandler() noexcept(false) {
+    if (needToRecover_) {
+        hipError_t err;
+        HIP_ERROR_FUNC(hipCtxSetCurrent, err, original_);
+    }
+    delete placedContext_;
+}
+
+void ContextCallback(void *userData) {
+    auto *ptr = static_cast<std::atomic<rocblas_handle> *>(userData);
```

Comment:
rocsolver_handle?

Reply:
Usage of rocsolver_handle is [deprecated](https://rocsolver.readthedocs.io/en/rocm-5.2.1/api_deprecated.html?highlight=rocsolver_handle#rocsolver-handle:~:text=Deprecated%20since%20version%203.5%3A%20Use%20rocblas_handle) and as suggested here, rocblas_handle is used instead.